pkg/controller/podgc/gc_controller.go

```diff
@@ -141,23 +157,36 @@ func (gcc *PodGCController) gcTerminated(pods []*v1.Pod) {
 }
 
 // gcOrphaned deletes pods that are bound to nodes that don't exist.
-func (gcc *PodGCController) gcOrphaned(pods []*v1.Pod) {
+func (gcc *PodGCController) gcOrphaned(pods []*v1.Pod, nodes []*v1.Node) {
 	klog.V(4).Infof("GC'ing orphaned")
-	// We want to get list of Nodes from the etcd, to make sure that it's as fresh as possible.
-	nodes, err := gcc.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
-	if err != nil {
-		return
-	}
-	nodeNames := sets.NewString()
-	for i := range nodes.Items {
-		nodeNames.Insert(nodes.Items[i].Name)
+	existingNodeNames := sets.NewString()
+	for _, node := range nodes {
+		existingNodeNames.Insert(node.Name)
 	}
-
+	// Add newly found unknown nodes to quarantine
 	for _, pod := range pods {
-		if pod.Spec.NodeName == "" {
+		if pod.Spec.NodeName != "" && !existingNodeNames.Has(pod.Spec.NodeName) {
+			gcc.nodeQueue.EnqueueIfNew(pod.Spec.NodeName, quarantineTime)
+		}
+	}
+	// Check if nodes are still missing after quarantine period
+	deletedNodesNames := sets.NewString()
+	for _, nodeName := range gcc.nodeQueue.GetWork() {
+		if existingNodeNames.Has(nodeName) {
 			continue
 		}
-		if nodeNames.Has(pod.Spec.NodeName) {
+		exists, err := gcc.checkIfNodeExists(nodeName)
+		if err != nil {
+			klog.Errorf("Error while getting node %q: %v", nodeName, err)
+			continue
```

Comment:
Error handling is broken here - if we couldn't get node due to some transient error, we should requeue it retry or sth like that.

Reply:
it will be requeued like any new node

Reply:
How?

Reply:
There would be still orphaned pods so their node would end up in the queue again.

Reply:
I need to think about it - but assuming you're right, it at least requires a comment why you're not taking any action on error here.

Reply:
Done